### DISCUSSION TITLE
Returning Specific Non MST Type From Map in Typescript

### DISCUSSION THREAD

**Discussion by mashaalmemon** [mobxjs/mobx-state-tree#1768] [Q&A]
I'm considering using Mobx-State-Tree in an app where all of our types are already defined using typescript. 

For example I have types already defined for different data entities, and these were define prior to considering Mobx-State-Tree for our global store:

```
/**
 * User record used in entire system other than Mobx-State-Tree
 */
interface IUserRecord {
  tsid: Moment;
  uuid: string;
  name: string;
}
```

I understand that to use Mobx-State-Tree to model my entities I'll need to use mobx `types`:

```
/**
 * User record for Mobx-State-Tree
 */
export const User = types.model({
    tsid: momentDate, // custom Mobx-State-Tree type 
    uuid: types.string,
    name: types.string,
    // make id type optional so that can pass in model without it when passing to action, 
    // but then have preProcessSnapshot fill id in
    // @ts-ignore
    id: types.optional(types.identifier, ()=>{ throw Error('id not defined.'); return ""}),
}).preProcessSnapshot(
    ((snapshot)=> {
        // augment all users with id
        const id = buildCompositeIdUser(snapshot.name, snapshot.uuid);
        return {...snapshot, id};
    })
)
```

And I intend to use a map to take advantage of references to keep my data normalized:

```
/**
 * Root of the MST, stored here.
 */
export const Store = types.model({
    users: types.map(User)
}).actions(...)
```

When observing changes to the `users` map in react, and getting back an object, I understand from a Typescript typing perspective, I will get back a `Instance<typeof User>` typed object back.

My main question is,  is there a way I can set it such that I get a `IUserRecord` typed object back instead. For my purposes they will be almost equivalent in terms of the data it holds (other than the `id` field). Yes I can do explicit casting in typescript everytime I observe the `users` map, but I'm hoping there is a better way.

**Comment by Ridermansb** [comment]
I posted an similar issue..

https://github.com/mobxjs/mobx-state-tree/issues/1840
